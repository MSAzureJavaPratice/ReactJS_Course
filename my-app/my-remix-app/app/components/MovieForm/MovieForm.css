.movie-form {
    display: flex;
    flex-direction: column;
    gap: 15px;
}

.movie-form label {
    display: flex;
    flex-direction: column;
    font-weight: bold;
}

.movie-form button {
    padding: 10px;
    background-color: red;
    color: white;
    border: none;
    border-radius: 4px;
    cursor: pointer;
}