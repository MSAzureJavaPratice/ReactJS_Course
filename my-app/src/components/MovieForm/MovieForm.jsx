import React, { useEffect, useState } from "react";
import { useForm, Controller } from "react-hook-form";
import "./MovieForm.css";

const availableGenres = [
  "Action",
  "Crime",
  "Drama",
  "Sci-Fi",
  "Thriller",
  "Fantasy",
  "Adventure",
  "Romance",
];

const MovieForm = ({ initialMovieInfo = {}, onSubmit }) => {
  const [movieId, setMovieId] = useState(initialMovieInfo.id || "");

  const {
    register,
    handleSubmit,
    control,
    reset,
    formState: { errors },
  } = useForm({
    defaultValues: {
      title: initialMovieInfo.title || "",
      release_date: initialMovieInfo.release_date || "",
      genres: initialMovieInfo.genres || [],
      runtime: initialMovieInfo.runtime || "",
      overview: initialMovieInfo.overview || "",
      vote_average: initialMovieInfo.vote_average || "",
      poster_path: initialMovieInfo.poster_path || "",
    },
  });

  useEffect(() => {
    // Prevent unnecessary resets by checking for actual changes
    if (Object.keys(initialMovieInfo).length > 0) {
      console.log("Resetting form with:", initialMovieInfo);

      reset({
        title: initialMovieInfo.title || "",
        release_date: initialMovieInfo.release_date || "",
        genres: initialMovieInfo.genres || [],
        runtime: initialMovieInfo.runtime || "",
        overview: initialMovieInfo.overview || "",
        vote_average: initialMovieInfo.vote_average || "",
        poster_path: initialMovieInfo.poster_path || "",
      });

      setMovieId(initialMovieInfo.id || "");
    }
  }, [initialMovieInfo, reset]);

  const submitHandler = (data) => {
    console.log('data submission: ' + JSON.stringify(data));
    onSubmit({
      ...data,
      id: movieId, // Include the movie ID in the submitted data
      runtime: parseInt(data.runtime, 10),
      vote_average: parseFloat(data.vote_average),
    });
  };

  return (
    <form
      className="movie-form"
      onSubmit={handleSubmit(submitHandler)}
      data-testid="movie-form"
    >
      <div>
        <label>
          Title:
          <input
            {...register("title", { required: "Title is required" })}
            data-testid="title"
          />
          {errors.title && (
            <span className="error">{errors.title.message}</span>
          )}
        </label>
      </div>
      <div>
        <label>
          Release Date:
          <input
            type="date"
            {...register("release_date", {
              required: "Release date is required",
            })}
            data-testid="release_date"
          />
          {errors.release_date && (
            <span className="error">{errors.release_date.message}</span>
          )}
        </label>
      </div>
      <div>
        <label>
          Genre:
          <Controller
            control={control}
            name="genres"
            rules={{ required: "At least one genre must be selected" }}
            render={({ field: { value, onChange } }) => (
              <select
                multiple
                value={value || []}
                onChange={(e) => {
                  const selectedOptions = Array.from(
                    e.target.selectedOptions,
                    (option) => option.value
                  );
                  onChange(selectedOptions);
                }}
                data-testid="genres"
              >
                {availableGenres.map((genre) => (
                  <option key={genre} value={genre}>
                    {genre}
                  </option>
                ))}
              </select>
            )}
          />
        </label>
        {errors.genres && (
          <span className="error">{errors.genres.message}</span>
        )}
      </div>
      <div>
        <label>
          Runtime (minutes):
          <input
            type="number"
            {...register("runtime", {
              required: "Runtime is required",
              min: { value: 1, message: "Runtime must be at least 1 minute" },
            })}
            data-testid="runtime"
          />
          {errors.runtime && (
            <span className="error">{errors.runtime.message}</span>
          )}
        </label>
      </div>
      <div>
        <label>
          Rating:
          <input
            type="number"
            step="0.1"
            {...register("vote_average", {
              required: "Rating is required",
              min: { value: 0, message: "Rating must be at least 0" },
              max: { value: 10, message: "Rating cannot exceed 10" },
            })}
            data-testid="vote_average"
          />
          {errors.vote_average && (
            <span className="error">{errors.vote_average.message}</span>
          )}
        </label>
      </div>
      <div>
        <label>
          Poster URL:
          <input
            {...register("poster_path", {
              required: "Poster URL is required",
              pattern: {
                value: /^(https?:\/\/.*\.(?:png|jpg|jpeg|gif|svg))$/i,
                message: "Enter a valid image URL",
              },
            })}
            data-testid="poster_path"
          />
          {errors.poster_path && (
            <span className="error">{errors.poster_path.message}</span>
          )}
        </label>
      </div>
      <div>
        <label>
          Overview:
          <textarea
            rows="5"
            {...register("overview", {
              required: "Overview is required",
              minLength: {
                value: 10,
                message: "Overview must be at least 10 characters",
              },
            })}
            data-testid="overview"
          />
          {errors.overview && (
            <span className="error">{errors.overview.message}</span>
          )}
        </label>
      </div>
      <button type="submit" data-testid="submit-button">
        Submit
      </button>
    </form>
  );
};

export default MovieForm;
