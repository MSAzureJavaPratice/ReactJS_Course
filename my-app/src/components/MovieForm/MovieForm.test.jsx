import { render, screen, fireEvent, act } from "@testing-library/react";
import MovieForm from "./MovieForm";

describe("MovieForm Component", () => {
  const mockOnSubmit = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("renders empty form when no initialMovieInfo is provided", () => {
    render(<MovieForm onSubmit={mockOnSubmit} />);

    expect(screen.getByTestId("title")).toHaveValue("");
    expect(screen.getByTestId("release_date")).toHaveValue("");
    expect(screen.getByTestId("genres")).toHaveValue([]);
  });

  test("renders form with initial values when initialMovieInfo is provided", () => {
    const initialMovieInfo = {
      title: "Inception",
      release_date: "2010-07-16",
      genres: ["Action", "Sci-Fi"],
      runtime: 148,
      overview: "A mind-bending thriller.",
      vote_average: 8.8,
      poster_path: "https://example.com/inception.jpg",
    };

    render(<MovieForm initialMovieInfo={initialMovieInfo} onSubmit={mockOnSubmit} />);

    expect(screen.getByTestId("title")).toHaveValue("Inception");
    expect(screen.getByTestId("release_date")).toHaveValue("2010-07-16");
    expect(screen.getByTestId("genres").selectedOptions.length).toBe(2);
    expect(screen.getByTestId("runtime")).toHaveValue(148);
    expect(screen.getByTestId("overview")).toHaveValue("A mind-bending thriller.");
    expect(screen.getByTestId("vote_average")).toHaveValue(8.8);
    expect(screen.getByTestId("poster_path")).toHaveValue("https://example.com/inception.jpg");
  });

  it("calls onSubmit with form data when submitted", async () => {
    const initialMovieInfo = {
      title: "Inception",
      release_date: "2010-07-16",
      genres: ["Sci-Fi", "Thriller"],
      runtime: 148,
      overview: "A mind-bending thriller.",
      vote_average: 8.8,
      poster_path: "https://example.com/inception.jpg",
    };

    render(<MovieForm initialMovieInfo={initialMovieInfo} onSubmit={mockOnSubmit} />);

    // Fill out required fields
    fireEvent.change(screen.getByTestId("title"), {
      target: { value: "Inception" },
    });
    fireEvent.change(screen.getByTestId("runtime"), {
      target: { value: "148" },
    });

    // Wrap form submission in `act()`
    await act(async () => {
      fireEvent.submit(screen.getByTestId("movie-form"));
    });

    // Ensure onSubmit is called with the correct data
    expect(mockOnSubmit).toHaveBeenCalledWith(
      expect.objectContaining({
        title: "Inception",
        runtime: 148,
      })
    );
  });

  test("matches snapshot", () => {
    const { asFragment } = render(<MovieForm onSubmit={mockOnSubmit} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
