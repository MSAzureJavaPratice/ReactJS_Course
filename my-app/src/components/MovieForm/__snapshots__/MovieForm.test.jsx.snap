// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm Component matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="movie-form"
    data-testid="movie-form"
  >
    <div>
      <label>
        Title:
        <input
          data-testid="title"
          name="title"
        />
      </label>
    </div>
    <div>
      <label>
        Release Date:
        <input
          data-testid="release_date"
          name="release_date"
          type="date"
        />
      </label>
    </div>
    <div>
      <label>
        Genre:
        <select
          data-testid="genres"
          multiple=""
        >
          <option
            value="Action"
          >
            Action
          </option>
          <option
            value="Crime"
          >
            Crime
          </option>
          <option
            value="Drama"
          >
            Drama
          </option>
          <option
            value="Sci-Fi"
          >
            Sci-Fi
          </option>
          <option
            value="Thriller"
          >
            Thriller
          </option>
          <option
            value="Fantasy"
          >
            Fantasy
          </option>
          <option
            value="Adventure"
          >
            Adventure
          </option>
          <option
            value="Romance"
          >
            Romance
          </option>
        </select>
      </label>
    </div>
    <div>
      <label>
        Runtime (minutes):
        <input
          data-testid="runtime"
          name="runtime"
          type="number"
        />
      </label>
    </div>
    <div>
      <label>
        Rating:
        <input
          data-testid="vote_average"
          name="vote_average"
          step="0.1"
          type="number"
        />
      </label>
    </div>
    <div>
      <label>
        Poster URL:
        <input
          data-testid="poster_path"
          name="poster_path"
        />
      </label>
    </div>
    <div>
      <label>
        Overview:
        <textarea
          data-testid="overview"
          name="overview"
          rows="5"
        />
      </label>
    </div>
    <button
      data-testid="submit-button"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
